R/simulate.r

```diff
@@ -55,192 +40,404 @@
+                                    stat_latest = stat_track,
+                                    n_offspring = n_offspring
+                                    )
+
+    ## only continue to simulate chains that offspring and aren't of
+    ## chain_stat_max size/length
+    sim <- which(n_offspring > 0 & stat_track < chain_stat_max)
+  }
+
+  stat_track[stat_track >= chain_stat_max] <- Inf
+
+  structure(
+    stat_track,
+    chain_type = "chains_vec",
+    chains = nchains,
+    class = c("epichains", class(stat_track))
+  )
+}
+
+#' Simulate a tree of infections from an initial susceptible population
+#' with initial immunity
+#'
+#' @param pop The susceptible population.
+#' @param offspring_sampler Offspring distribution sampler: a character string
+#' corresponding to the R distribution function. Currently only "pois" &
+#' "nbinom" are supported. Internally truncated distributions are used to
+#' avoid infecting more people than susceptibles available.
+#' @param mean_offspring The average number of secondary cases for each case.
+#' Same as R0.
+#' @param disp_offspring The dispersion parameter of the number of
+#' secondary cases. Ignored if \code{offspring == "pois"}. Must be > 1 to
+#' avoid division by 0 when calculating the size. See details and
+#'  \code{?rnbinom} for details on the parameterisation in Ecology.
+#' @param serial_sampler The serial interval. A function that takes one
+#' parameter (`n`), the number of serial intervals to randomly sample. Value
+#' must be >= 0.
+#' @param initial_immune The number of initial immunes in the population.
+#' @param t0 Start time; Defaults to 0.
+#' @param tf End time; Defaults to `Inf`.
+#' @return a data frame with columns `time`, `id` (a unique ID for each
+#' individual element of the chain), `ancestor` (the ID of the ancestor
+#' of each element), and `generation`.
+#' @details
+#'
+#' # Offspring models
+#'
+#' The poisson model is parametrised so that:
+#'
+#' lamda = mean_offspring * pop - initial_immune / pop
+#'
+#' The negative binomial model is parametrised as:
+#'
+#' mu = mean_offspring * pop - initial immune / pop, and
+#' size = mu / (disp_offspring - 1). This is why disp_offspring must be greater
+#' than 1.
+#'
+#' simulate_tree_from_pop() has a couple of key different from simulate_tree():
+#'  * the maximal chain statistic is limited by `pop` instead of
+#'  `chain_stat_max` (in `simulate_tree()`),
+#'  * it can only handle implemented offspring distributions ("pois" and
+#' "nbinom").
+#' @author Flavio Finger
+#' @author James M. Azam
+#' @examples
+#' # Simulate with poisson offspring
+#' simulate_tree_from_pop(pop = 100, offspring_sampler = "pois",
+#' mean_offspring = 0.5, serial_sampler = function(x) 3)
+#'
+#' # Simulate with negative binomial offspring
+#' simulate_tree_from_pop(pop = 100, offspring_sampler = "nbinom",
+#' mean_offspring = 0.5, disp_offspring = 1.1, serial_sampler = function(x) 3)
+#' @export
+simulate_tree_from_pop <- function(pop,
```

Comment:
Is the plan to eventially merge this into `simulate_tree` with a `pop` option? I think that might make sense if it doesn't make that function too complex.